### DISCUSSION TITLE
Arduino UNO R4 WiFi OTA update

### DISCUSSION THREAD

**Discussion by Alessio6976** [JAndrassy/ArduinoOTA#258] [Q&A]
Hi everyone,

is there a way to update the firmware over the air for an Arduino UNO R4 WiFi, storing the new firmware on github? I saw some example, but nothing work.

I followed some tutorial, but in the "update.h" include there are many includes missing, like "md5.h"...."esp_spi_flash.h"...and every include I fix, it asks for another missing include... So I'm aware that is not possible.

I'm searching for a simple skecth that connects to wifi, and ping to a given url (I've just uploaded a new version of a sample firmware on github) to download and than upgrade the firmware:

My example (not working):

#include <WiFiS3.h>
#include <SPIFFS.h>
#include "Update.h"
#include <WiFiClientSecure.h>

// Definisci le credenziali WiFi
const char* ssid = "IlTuoSSID"; // Inserisci il tuo SSID
const char* password = "LaTuaPassword"; // Inserisci la tua password

// Definisci i dettagli del server e il percorso del file
#define HOST "raw.githubusercontent.com"
#define PATH "/ilTuoNomeUtente/ilTuoRepository/main/firmware.bin" // Sostituisci con il tuo percorso
#define PORT 443

// Definisci il nome del file per il firmware scaricato
#define FILE_NAME "/firmware.bin"

void setup() {
  Serial.begin(115200);

  // Inizializza SPIFFS
  if (!SPIFFS.begin(true)) {
    Serial.println("SPIFFS Mount Failed");
    return;
  }

  // Connettiti al WiFi
  connectToWiFi();

  // Scarica il firmware dal server
  getFileFromServer();

  // Esegui l'aggiornamento OTA
  performOTAUpdateFromSPIFFS();
}

void loop() {
  // Nessuna azione nel loop
}

void connectToWiFi() {
  // Inizia la connessione al WiFi usando SSID e password forniti
  WiFi.begin(ssid, password);

  // Mostra il progresso della connessione
  Serial.print("Connecting to WiFi");
  
  // Attendi finché non è connesso al WiFi
  while (WiFi.status() != WL_CONNECTED) {
    delay(500);
    Serial.print(".");
  }
  
  // Messaggio di conferma quando il WiFi è connesso
  Serial.println("WiFi connected");
}

void getFileFromServer() {
  WiFiClientSecure client;
  client.setInsecure(); // Imposta il client per permettere connessioni insicure

  if (client.connect(HOST, PORT)) { // Connettiti al server
    Serial.println("Connected to server");
    client.print("GET " + String(PATH) + " HTTP/1.1\r\n"); // Invia richiesta GET HTTP
    client.print("Host: " + String(HOST) + "\r\n"); // Specifica l'host
    client.println("Connection: close\r\n"); // Chiudi la connessione dopo la risposta
    client.println(); // Invia una riga vuota per indicare la fine delle intestazioni della richiesta

    File file = SPIFFS.open(FILE_NAME, FILE_WRITE); // Apri il file in SPIFFS per la scrittura
    if (!file) {
      Serial.println("Failed to open file for writing");
      return;
    }

    bool endOfHeaders = false;
    String headers = "";
    String http_response_code = "error";
    const size_t bufferSize = 1024; // Dimensione del buffer per la lettura dei dati
    uint8_t buffer[bufferSize];

    // Loop per leggere le intestazioni della risposta HTTP
    while (client.connected() && !endOfHeaders) {
      if (client.available()) {
        char c = client.read();
        headers += c;
        if (headers.startsWith("HTTP/1.1")) {
          http_response_code = headers.substring(9, 12);
        }
        if (headers.endsWith("\r\n\r\n")) { // Verifica la fine delle intestazioni
          endOfHeaders = true;
        }
      }
    }

    Serial.println("HTTP response code: " + http_response_code); // Stampa le intestazioni ricevute

    // Loop per leggere e scrivere i dati grezzi nel file
    while (client.connected()) {
      if (client.available()) {
        size_t bytesRead = client.readBytes(buffer, bufferSize);
        file.write(buffer, bytesRead); // Scrivi i dati nel file
      }
    }
    file.close(); // Chiudi il file
    client.stop(); // Chiudi la connessione del client
    Serial.println("File saved successfully");
  }
  else {
    Serial.println("Failed to connect to server");
  }
}

void performOTAUpdateFromSPIFFS() {
  // Apri il file del firmware in SPIFFS per la lettura
  File file = SPIFFS.open(FILE_NAME, FILE_READ);
  if (!file) {
    Serial.println("Failed to open file for reading");
    return;
  }

  Serial.println("Starting update..");
  size_t fileSize = file.size(); // Ottieni la dimensione del file
  Serial.println(fileSize);

  // Inizia il processo di aggiornamento OTA con la dimensione specificata e la destinazione flash
  if (!Update.begin(fileSize, U_FLASH)) {
    Serial.println("Cannot do the update");
    return;
  }

  // Scrivi i dati del firmware dal file all'aggiornamento OTA
  Update.writeStream(file);

  // Completa il processo di aggiornamento OTA
  if (Update.end()) {
    Serial.println("Successful update");
  }
  else {
    Serial.println("Error Occurred:" + String(Update.getError()));
    return;
  }

  file.close(); // Chiudi il file
  Serial.println("Reset in 4 seconds....");
  delay(4000);
  ESP.restart(); // Riavvia l'ESP32 per applicare l'aggiornamento
}

**Comment by JAndrassy** [comment]
the sketch you posted is not for this library and not for Uno R4.
see the Advanced examples of this library

**Comment by Alessio6976** [comment]
Thank you JAndrassy, your advanced example works perfectly. I only made a little update for UNO R4 WiFi, but now it's perfect!

Thank you
You saved me!

**Reply by mikelorfe** [reply]
Hi, good day. Could you share the code that is working for you? I can't get mine to work.

**Reply by JAndrassy** [reply]
@mikelorfe try the WiFi101OTA example. just change WiFiNINA include to WiFiS3

**Reply by mikelorfe** [reply]
loading library from C:\Users\mikel\AppData\Local\Arduino15\libraries\OLD: invalid library: no header files foundloading library from c:\Users\mikel\Documents\Arduino\libraries\OLD: invalid library: no header files found
FQBN: arduino:renesas_uno:unor4wifi
Using board 'unor4wifi' from platform in folder: C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2
Using core 'arduino' from platform in folder: C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2

Detecting libraries used...
C:\Users\mikel\AppData\Local\arduino15\packages\arduino\tools\arm-none-eabi-gcc\7-2017q4/bin/arm-none-eabi-g++ -c -w -Os -g3 -fno-use-cxa-atexit -fno-rtti -fno-exceptions -nostdlib -DF_CPU=48000000 -DNO_USB -DBACKTRACE_SUPPORT -DARDUINO_UNOR4_WIFI -std=gnu++17 -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsigned-char -ffunction-sections -fdata-sections -fmessage-length=0 -fno-builtin -w -x c++ -E -CC -DARDUINO=10607 -DPROJECT_NAME="C:\Users\mikel\AppData\Local\arduino\sketches\EBB485858B4AA01988EFE00EF84CF46F/WiFi101_OTA.ino" -DARDUINO_UNOWIFIR4 -DARDUINO_ARCH_RENESAS_UNO -DARDUINO_ARCH_RENESAS -DARDUINO_FSP -D_XOPEN_SOURCE=700 -mthumb @C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/defines.txt -DCFG_TUSB_MCU=OPT_MCU_RAXXX -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/tinyusb -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated-avr-comp -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4 -iprefixC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2 @C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/includes.txt C:\Users\mikel\AppData\Local\arduino\sketches\EBB485858B4AA01988EFE00EF84CF46F\sketch\WiFi101_OTA.ino.cpp -o nul
Alternatives for ArduinoOTA.h: [ArduinoOTA@3.2.0]
ResolveLibrary(ArduinoOTA.h)
  -> candidates: [ArduinoOTA@3.2.0]
C:\Users\mikel\AppData\Local\arduino15\packages\arduino\tools\arm-none-eabi-gcc\7-2017q4/bin/arm-none-eabi-g++ -c -w -Os -g3 -fno-use-cxa-atexit -fno-rtti -fno-exceptions -nostdlib -DF_CPU=48000000 -DNO_USB -DBACKTRACE_SUPPORT -DARDUINO_UNOR4_WIFI -std=gnu++17 -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsigned-char -ffunction-sections -fdata-sections -fmessage-length=0 -fno-builtin -w -x c++ -E -CC -DARDUINO=10607 -DPROJECT_NAME="C:\Users\mikel\AppData\Local\arduino\sketches\EBB485858B4AA01988EFE00EF84CF46F/WiFi101_OTA.ino" -DARDUINO_UNOWIFIR4 -DARDUINO_ARCH_RENESAS_UNO -DARDUINO_ARCH_RENESAS -DARDUINO_FSP -D_XOPEN_SOURCE=700 -mthumb @C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/defines.txt -DCFG_TUSB_MCU=OPT_MCU_RAXXX -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/tinyusb -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated-avr-comp -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4 -Ic:\Users\mikel\Documents\Arduino\libraries\ArduinoOTA\src -iprefixC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2 @C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/includes.txt C:\Users\mikel\AppData\Local\arduino\sketches\EBB485858B4AA01988EFE00EF84CF46F\sketch\WiFi101_OTA.ino.cpp -o nul
Alternatives for Network.h: [Networking@3.2.0]
ResolveLibrary(Network.h)
  -> candidates: [Networking@3.2.0]
C:\Users\mikel\AppData\Local\arduino15\packages\arduino\tools\arm-none-eabi-gcc\7-2017q4/bin/arm-none-eabi-g++ -c -w -Os -g3 -fno-use-cxa-atexit -fno-rtti -fno-exceptions -nostdlib -DF_CPU=48000000 -DNO_USB -DBACKTRACE_SUPPORT -DARDUINO_UNOR4_WIFI -std=gnu++17 -mcpu=cortex-m4 -mfloat-abi=hard -mfpu=fpv4-sp-d16 -fsigned-char -ffunction-sections -fdata-sections -fmessage-length=0 -fno-builtin -w -x c++ -E -CC -DARDUINO=10607 -DPROJECT_NAME="C:\Users\mikel\AppData\Local\arduino\sketches\EBB485858B4AA01988EFE00EF84CF46F/WiFi101_OTA.ino" -DARDUINO_UNOWIFIR4 -DARDUINO_ARCH_RENESAS_UNO -DARDUINO_ARCH_RENESAS -DARDUINO_FSP -D_XOPEN_SOURCE=700 -mthumb @C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/defines.txt -DCFG_TUSB_MCU=OPT_MCU_RAXXX -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/tinyusb -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino/api/deprecated-avr-comp -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\cores\arduino -IC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4 -Ic:\Users\mikel\Documents\Arduino\libraries\ArduinoOTA\src -Ic:\Users\mikel\Documents\Arduino\libraries\Network\src -iprefixC:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2 @C:\Users\mikel\AppData\Local\arduino15\packages\arduino\hardware\renesas_uno\1.3.2\variants\UNOWIFIR4/includes.txt C:\Users\mikel\AppData\Local\arduino\sketches\EBB485858B4AA01988EFE00EF84CF46F\sketch\WiFi101_OTA.ino.cpp -o nul
In file included from c:\Users\mikel\Documents\Arduino\libraries\Network\src/Network.h:8:0,
                 from c:\Users\mikel\Documents\Arduino\libraries\ArduinoOTA\src/ArduinoOTA.h:18,
                 from C:\Users\mikel\AppData\Local\Temp\.arduinoIDE-unsaved2025120-42392-byktjz.bbjj\WiFi101_OTA\WiFi101_OTA.ino:1:
c:\Users\mikel\Documents\Arduino\libraries\Network\src/NetworkInterface.h:8:10: fatal error: esp_netif_types.h: No such file or directory
 #include "esp_netif_types.h"
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
Alternatives for esp_netif_types.h: []
ResolveLibrary(esp_netif_types.h)
  -> candidates: []
exit status 1

Compilation error: exit status 1

**Reply by mikelorfe** [reply]
And thanks for the prompt response

**Reply by JAndrassy** [reply]
delete c:\Users\mikel\Documents\Arduino\libraries\Network

**Reply by mikelorfe** [reply]
I don't know what's happening, but when I run the example and switch from wifinina to wifiS3, it says the library doesn't exist, and I have to add it manually. The same thing happens with the SPI and arduinoOTA libraries. If I try to compile now, it says the network.h library doesn't exist.

**Reply by JAndrassy** [reply]
try an example of the WiFiS3 library

**Reply by mikelorfe** [reply]
Okay, I tried it, and it compiles and works.

**Reply by JAndrassy** [reply]
did you install ArduinoOTA library with Library Manager?

**Reply by mikelorfe** [reply]
yes i did

**Reply by JAndrassy** [reply]
this library is at version 1.1.0 and your compile output shows 3.2.0.
why do you have esp32 platform libraries in the libraries folder?

**Reply by mikelorfe** [reply]
I tried several things I saw on Arduino forums. I uninstalled and reinstalled the ArduinoOTA library, but it still fails.

**Reply by mikelorfe** [reply]
i removed all the libraries and its working now thanks for all
